File: app/src/main/java/com/example/taskapp/data/TaskRepository.java
package com.example.taskapp.data;

import java.util.List;
import retrofit2.Call;

public class TaskRepository {
    private final ApiService api;
    private final AuthManager authManager;

    public TaskRepository(ApiService api, AuthManager authManager) {
        this.api = api;
        this.authManager = authManager;
    }

    public Call<ApiResponse> register(AuthRequest auth) {
        return api.register(auth);
    }

    public Call<ApiResponse> login(AuthRequest auth) {
        return api.login(auth);
    }

    public Call<List<Task>> getTasks() {
        AuthRequest auth = authManager.getCredentials();
        if (auth == null) return null;
        return api.getTasks(auth);
    }

    public Call<Task> createTask(String title) {
        AuthRequest auth = authManager.getCredentials();
        if (auth == null) return null;
        return api.createTask(new TaskCreateRequest(auth.getUsername(), auth.getPassword(), title));
    }

    public Call<Task> updateTask(int id, boolean completed) {
        AuthRequest auth = authManager.getCredentials();
        if (auth == null) return null;
        return api.updateTask(id, new TaskUpdateRequest(auth.getUsername(), auth.getPassword(), null, completed));
    }

    public Call<Task> updateTaskTitle(int id, String title) {
        AuthRequest auth = authManager.getCredentials();
        if (auth == null) return null;
        return api.updateTask(id, new TaskUpdateRequest(auth.getUsername(), auth.getPassword(), title, null));
    }

    public Call<ApiResponse> deleteTask(int id) {
        AuthRequest auth = authManager.getCredentials();
        if (auth == null) return null;
        return api.deleteTask(id, new TaskDeleteRequest(auth.getUsername(), auth.getPassword()));
    }

    public void saveAuth(AuthRequest auth) { authManager.saveCredentials(auth); }
    public AuthRequest getAuth() { return authManager.getCredentials(); }
    public void clearAuth() { authManager.clear(); }
}

File: app/src/main/java/com/example/taskapp/data/TaskRepository.kt
package com.example.taskapp.data

import retrofit2.Response

class TaskRepository(private val api: ApiService, private val authManager: AuthManager) {

    suspend fun register(auth: AuthRequest): Response<ApiResponse> = api.register(auth)

    suspend fun getTasks(): Response<List<Task>> {
        val auth = authManager.getCredentials() ?: return Response.error(401, okhttp3.ResponseBody.create(null, "No credentials"))
        return api.getTasks(auth)
    }

    suspend fun createTask(title: String): Response<Task> {
        val auth = authManager.getCredentials() ?: return Response.error(401, okhttp3.ResponseBody.create(null, "No credentials"))
        return api.createTask(TaskCreateRequest(auth.username, auth.password, title))
    }

    suspend fun updateTask(id: Int, completed: Boolean): Response<Task> {
        val auth = authManager.getCredentials() ?: return Response.error(401, okhttp3.ResponseBody.create(null, "No credentials"))
        return api.updateTask(id, TaskUpdateRequest(auth.username, auth.password, completed = completed))
    }

    suspend fun updateTaskTitle(id: Int, title: String): Response<Task> {
        val auth = authManager.getCredentials() ?: return Response.error(401, okhttp3.ResponseBody.create(null, "No credentials"))
        return api.updateTask(id, TaskUpdateRequest(auth.username, auth.password, title = title))
    }

    suspend fun deleteTask(id: Int): Response<ApiResponse> {
        val auth = authManager.getCredentials() ?: return Response.error(401, okhttp3.ResponseBody.create(null, "No credentials"))
        return api.deleteTask(id, TaskDeleteRequest(auth.username, auth.password))
    }
    
    fun saveAuth(auth: AuthRequest) = authManager.saveCredentials(auth)
    fun getAuth() = authManager.getCredentials()
    fun clearAuth() = authManager.clear()
}
